### PR TITLE
Updating assembly version for oob and inbox assemblies

### PR DESCRIPTION
**Pull request by Anipik**
- Wrote a small code to increment the assembly version in directory.build.props by 1
- manually increment the shims version
- manually added a netcoreapp3.1 for inbox assemblies (Microsoft.private.corefx validation fails without this)
- some of the assembly version cant be updating because they require the same verison on full framework, so they needed to be revereted
- used UpdateRepoPackageIndex target to update the assembly-package dependency for shipping nuget packages

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <Project>
   <PropertyGroup>
     <!-- The .NET Core product branding version -->
-    <ProductVersion>3.0.0</ProductVersion>
+    <ProductVersion>3.1.0</ProductVersion>
```
Comment (mmitche) [inline eng/Versions.props]: You also need to update the prerelease label (to preview1) below

File: src/System.ComponentModel.Composition.Registration/Directory.Build.props [hunk 2]
```diff
@@ -1,7 +1,7 @@
-﻿<Project>
+<Project>
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
     <AssemblyVersion>4.0.0.0</AssemblyVersion>
```
Comment (safern) [inline src/System.ComponentModel.Composition.Registration/Directory.Build.props]: Shouldn't this be bumped and then the ref assembly leave it as 4.0.0.0 in order to match full framework?

File: src/System.Data.SqlClient/Directory.Build.props [hunk 3]
```diff
@@ -1,9 +1,9 @@
-﻿<Project>
+<Project>
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
     <!-- Must be kept in sync with pkg\Microsoft.Windows.Compatibility\Microsoft.Windows.Compatibility.pkgproj -->
     <PackageVersion>4.8.0</PackageVersion>
```
Comment (safern) [inline src/System.Data.SqlClient/Directory.Build.props]: Should this be bumped to 4.9.0? Since 4.8.0 was the 3.0 version of this package. @ericstj?

File: src/System.DirectoryServices.AccountManagement/Directory.Build.props [hunk 4]
```diff
@@ -1,4 +1,4 @@
-﻿<Project>
+<Project>
```
Comment (safern) [inline src/System.DirectoryServices.AccountManagement/Directory.Build.props]: Could you revert this changes? It seems like your code is changing characters in our files.

Comment (ericstj) [reply src/System.DirectoryServices.AccountManagement/Directory.Build.props]: Probably adding or removing the BOM.

File: src/System.Data.SqlClient/Directory.Build.props (hunk 3 above)
Comment (ericstj) [reply src/System.Data.SqlClient/Directory.Build.props]: 3.0 was 4.7.0 https://github.com/dotnet/corefx/blob/5f2c2771e077f98c097ecc1b9202bee22ef50aed/src/System.Data.SqlClient/Directory.Build.props#L5, this should be 4.8.0.

File: pkg/Microsoft.Private.CoreFx.UAP/Microsoft.Private.CoreFx.UAP.pkgproj [hunk 5]
```diff
@@ -1,6 +1,7 @@
 ﻿<Project DefaultTargets="Build">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), Directory.Build.props))\Directory.Build.props" />
   <PropertyGroup>
+    <SkipValidatePackage>true</SkipValidatePackage>
```
Comment (ericstj) [inline pkg/Microsoft.Private.CoreFx.UAP/Microsoft.Private.CoreFx.UAP.pkgproj]: This needs a comment and an issue.

File: src/System.ComponentModel.Composition.Registration/Directory.Build.props (hunk 2 above)
Comment (ericstj) [reply src/System.ComponentModel.Composition.Registration/Directory.Build.props]: Yes, in cases where we still unify with full framework and ship a package we *freeze* the ref and then keep incrementing the implementation.

File: src/System.Data.SqlClient/Directory.Build.props (hunk 3 above)
Comment (safern) [reply src/System.Data.SqlClient/Directory.Build.props]: Ah right, it confused me that it was not bumped in this change.

Comment (ericstj) [reply src/System.Data.SqlClient/Directory.Build.props]: No worries.  This sort of confusion will hopefully go away in 5.0.  Speaking of which: we should decide what to do with SqlClient in 5.0.  We may want to remove that similar to the Microsoft.Bcl packages.  Can we open an issue and tag sql client owners?

File: pkg/Microsoft.Private.CoreFx.UAP/Microsoft.Private.CoreFx.UAP.pkgproj (hunk 5 above)
Comment (ViktorHofer) [reply pkg/Microsoft.Private.CoreFx.UAP/Microsoft.Private.CoreFx.UAP.pkgproj]: I clarified offline, we should remove the transport package. You decide if you want to do it here or separately.

File: src/Microsoft.Bcl.AsyncInterfaces/Directory.Build.props [hunk 6]
```diff
@@ -1,9 +1,9 @@
 <Project>
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
-    <AssemblyVersion>1.0.0.0</AssemblyVersion>
+    <AssemblyVersion>1.0.1.0</AssemblyVersion>
```
Comment (ViktorHofer) [inline src/Microsoft.Bcl.AsyncInterfaces/Directory.Build.props]: We don't want to update this assembly as it will be deleted.

File: src/Microsoft.Bcl.HashCode/Directory.Build.props [hunk 7]
```diff
@@ -1,9 +1,9 @@
 <Project>
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
-    <AssemblyVersion>1.0.0.0</AssemblyVersion>
+    <AssemblyVersion>1.0.1.0</AssemblyVersion>
```
Comment (ViktorHofer) [inline src/Microsoft.Bcl.HashCode/Directory.Build.props]: same as above.

**Comment by ViktorHofer** [comment]
Can somebody please remind me of the rules for updating our assemblies? I see patch versions being updated here but I'm sure there is more to know.

**Comment by ericstj** [comment]
Package versions - A.B.C
A = our discretion / branding, we'll do this in 5.0
B = release
C = servicing

Assembly versions - A.B.C.D
A = our discretion / branding, we'll do this in 5.0
B = API / features
C = release
D = servicing

References are *special* since we cannot change the assembly version of a reference if that reference needs to bind to a runtime assembly that isn't part of the package we ship, so that's a big caveat here that makes things different.
Desktop references are also *special*, since they must match implementation in order to get the correct binding redirects, so we may be constrained for a netstandard reference due to above but have to ship a versioned reference for this reason.

In 5.0 we will stop manually managing versions where we can, so this policy should be less special to corefx.

File: pkg/Microsoft.Private.CoreFx.UAP/Microsoft.Private.CoreFx.UAP.pkgproj (hunk 5 above)
Comment (Anipik) [reply pkg/Microsoft.Private.CoreFx.UAP/Microsoft.Private.CoreFx.UAP.pkgproj]: I will remove it from 5.0 but do we have to remove it from 3.1 branch as well ?

Comment (ericstj) [reply pkg/Microsoft.Private.CoreFx.UAP/Microsoft.Private.CoreFx.UAP.pkgproj]: Yeah, let's remove it here too.

Comment (ViktorHofer) [reply pkg/Microsoft.Private.CoreFx.UAP/Microsoft.Private.CoreFx.UAP.pkgproj]: Removing it in the 3.1 branch as well is desired as that will be our LTSB.

**Comment by Anipik** [comment]
```
System.Resources.Extensions.Tests.PreserializedResourceWriterTests.EmbeddedResourcesAreUpToDate [FAIL]
      Assert.Equal() Failure
      Expected: Byte[] [206, 202, 239, 190, 1, ...]
      Actual:   Byte[] [206, 202, 239, 190, 1, ...]
      Stack Trace:
        /_/src/System.Resources.Extensions/tests/BinaryResourceWriterUnitTest.cs(513,0): at System.Resources.Extensions.Tests.PreserializedResourceWriterTests.EmbeddedResourcesAreUpToDate()
```
@ericstj this is the test failing on all platforms

**Comment by ericstj** [comment]
This test is failing to "observe" that the generated resources are not in sync with what would be generated by the current build.  It's there to remind you to update the resources.  Once you do a full build locally, you need to build the test project, then run dotnet msbuild /t:GenerateTestResourcesFile.  This will modify the checked in `.resources` file.  Once you do that, you need to rebuild the test so that it contains the updated resources.  After that the test should succeed.  Can you try this?

**Comment by ericstj** [comment]
Can you update the s.r.e test resource generation target?

**Comment by Anipik** [comment]
> Can you update the s.r.e test resource generation target?

I am still trying to find a way to pass the runtimeConfig parameter to the remoteexecutor.dll

currently the command is expected to be like this ```Console.Error.WriteLine("Usage: {0} assemblyName typeName methodName exceptionFile [additionalArgs]", typeof(Program).GetTypeInfo().Assembly.GetName().Name);```

**Comment by ericstj** [comment]
I can help with that

**Comment by ViktorHofer** [comment]
> I am still trying to find a way to pass the runtimeConfig parameter to the remoteexecutor.dll

What are you trying to accomplish?

**Comment by Anipik** [comment]
@ericstj @ViktorHofer @safern  the ci is finally green, can you review this one ?

File: pkg/test/frameworkSettings/netcoreapp3.1/settings.targets [hunk 8]
```diff
@@ -3,6 +3,8 @@
     <TargetFramework>netcoreapp3.1</TargetFramework>
     <NETCoreAppMaximumVersion>3.1</NETCoreAppMaximumVersion>
     <RuntimeFrameworkVersion>$(MicrosoftNETCoreAppPackageVersion)</RuntimeFrameworkVersion>
+    <!-- This should be enabled once we have updating assembly versions in our targeting pack. -->
```
Comment (ericstj) [inline pkg/test/frameworkSettings/netcoreapp3.1/settings.targets]: To be clear, this should be enabled once we can ingest a netcoreapp3.1 ref-pack that has the change from this PR.

File: src/System.ComponentModel.Composition.Registration/ref/System.ComponentModel.Composition.Registration.csproj [hunk 9]
```diff
@@ -1,5 +1,9 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
+    <!-- this assembly is inbox in desktop, do not version it unless you
```
Comment (ericstj) [inline src/System.ComponentModel.Composition.Registration/ref/System.ComponentModel.Composition.Registration.csproj]: Technically since we're building for `netstandard2.1` which doesn't apply to desktop we can version it.  Did you hit an error that forced you to freeze this?  I see https://github.com/dotnet/corefx/blob/80adc8b87163dfe77d5b4262ea7970cc2de87f46/src/System.ComponentModel.Composition.Registration/pkg/System.ComponentModel.Composition.Registration.pkgproj#L5, we could just remove the `net45` there.

File: src/System.Memory/Directory.Build.props [hunk 10]
```diff
@@ -2,10 +2,10 @@
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
     <!-- System.Memory has forwarded types into the runtime on netcoreapp/uap
-         It must win over assemblies versioned at 4.0.* -->
-    <AssemblyVersion>4.2.0.0</AssemblyVersion>
+It must win over assemblies versioned at 4.0.* -->
```
Comment (ericstj) [inline src/System.Memory/Directory.Build.props]: nit: whitespace.  Also we can just remove this entire comment, it was only relevant when we were shipping the package.

**Review by ericstj — APPROVED** [review]
just a few small points of feedback.  I reviewed all the special cases.

File: src/System.ComponentModel.Composition.Registration/pkg/System.ComponentModel.Composition.Registration.pkgproj [hunk 11]
```diff
@@ -2,12 +2,9 @@
     <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), Directory.Build.props))\Directory.Build.props" />
     <ItemGroup>
     <ProjectReference Include="..\ref\System.ComponentModel.Composition.Registration.csproj">
-      <SupportedFramework>netcoreapp3.0;net45</SupportedFramework>
+      <SupportedFramework>netcoreapp3.0</SupportedFramework>
     </ProjectReference>
     <ProjectReference Include="..\src\System.ComponentModel.Composition.Registration.csproj" />
-    <InboxOnTargetFramework Include="net45">
```
Comment (Anipik) [inline src/System.ComponentModel.Composition.Registration/pkg/System.ComponentModel.Composition.Registration.pkgproj]: I think we needed to remove this as well in order to remove ```net45```. 
Does it look okay to you @ericstj ?

Comment (ericstj) [reply src/System.ComponentModel.Composition.Registration/pkg/System.ComponentModel.Composition.Registration.pkgproj]: Can you leave it in and instead declare support for net45 as follows:
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/package-projects.md#applicability-validation
```
<SupportedFramework Include="net45" Version="4.0.0.0" />
```

Comment (ericstj) [reply src/System.ComponentModel.Composition.Registration/pkg/System.ComponentModel.Composition.Registration.pkgproj]: FWIW this won't be needed once we can harvest from a previous release.  Then we'll assert the support based on the harvested package.

**Review by ericstj — APPROVED** [review]
